From 5881ad19bedd8522dc86795902b081aa4ed2c3a1 Mon Sep 17 00:00:00 2001
From: Meng Li <Meng.Li@windriver.com>
Date: Tue, 11 Apr 2023 18:30:54 +0800
Subject: [PATCH] s32: env: support for booting up fitImage

Add environment variable to support for booting up fitImage
by default.

Upstream-Status: Inappropriate [WR Linux specific]

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 include/configs/s32-cc.h | 39 +++++++++++++++++++++++++++++++++++++--
 1 file changed, 37 insertions(+), 2 deletions(-)

diff --git a/include/configs/s32-cc.h b/include/configs/s32-cc.h
index 107988609e..c405a2ced4 100644
--- a/include/configs/s32-cc.h
+++ b/include/configs/s32-cc.h
@@ -44,6 +44,7 @@
 #define S32CC_BOOT_SCR_ADDR		0x82F00000
 #define S32CC_FDT_ADDR			0x83000000
 #define S32CC_RAMDISK_ADDR		0x90000000
+#define S32CC_FIT_ADDR                  0xC4000000
 
 #define MMC_PART_FAT			2
 #define MMC_PART_EXT			3
@@ -126,8 +127,12 @@
 		" root=/dev/ram rw" LINUX_EARLY_CONSOLE \
 		LINUX_LOG_DISABLE EXTRA_BOOT_ARGS ";"\
 		"setenv flashsize " __stringify(FSL_QSPI_FLASH_SIZE) ";\0" \
+	"fitimage=fitImage\0" \
+	"fitimage_addr=" __stringify(S32CC_FIT_ADDR) "\0" \
+	"fit_config_header=conf-\0" \
 	"ipaddr=" S32CC_IPADDR "\0"\
 	"loadtftpfdt=tftp ${fdt_addr} ${fdt_file};\0" \
+	"loadfitimage=fatload mmc ${mmcdev}:${mmcpart} ${fitimage_addr} ${fitimage}\0" \
 	"loadtftpimage=tftp ${loadaddr} ${image};\0" \
 	"netargs=setenv bootargs console=${console},${baudrate} " \
 		"root=/dev/nfs " \
@@ -183,7 +188,13 @@
 		"then " \
 			"run mmcboot; " \
 		"else " \
-			"run netboot; " \
+			"if run loadfitimage; "\
+			"then " \
+				"run mmcargs; " \
+				"bootm ${fitimage_addr}#${fit_config_header}${fdt_file}; " \
+			"else " \
+				"run netboot; " \
+			"fi; " \
 		"fi; " \
 	"else " \
 		"run netboot;" \
@@ -238,7 +249,23 @@
 #    endif
 #    define CONFIG_BOOTCOMMAND \
 	PRECONFIG_BOOTCOMMAND \
-	"run flashboot;"
+	"run flashboot;" \
+	"if test $? -ne 0; then " \
+		"mmc dev ${mmcdev}; " \
+		"if mmc rescan; " \
+		"then " \
+			"if run loadimage; "\
+			"then " \
+				"run mmcboot; " \
+			"else " \
+				"if run loadfitimage; "\
+				"then " \
+					"run mmcargs; " \
+					"bootm ${fitimage_addr}#${fit_config_header}${fdt_file}; " \
+				"fi; " \
+			"fi; " \
+		"fi;" \
+	"fi"
 #  elif defined(CONFIG_SD_BOOT)
 #    if defined(CONFIG_XEN_SUPPORT)
 #      define CONFIG_BOOTCOMMAND XEN_BOOTCMD
@@ -266,6 +293,14 @@
 		"if run loadimage; "\
 		"then " \
 			"run mmcboot; " \
+		"else " \
+			"if run loadfitimage; "\
+			"then " \
+				"run mmcargs; " \
+				"bootm ${fitimage_addr}#${fit_config_header}${fdt_file}; " \
+			"else " \
+				"run netboot; " \
+			"fi; " \
 		"fi; " \
 	"fi"
 #    endif
-- 
2.34.1

